### PR TITLE
added capability to set profile image

### PR DESCRIPTION
**Pull request by dan-nyanko**
This accomplished by [react-native-image-picker](https://github.com/marcshilling/react-native-image-picker)

image
image

### CODE REVIEW

**Comment by sophiabersani** [comment]
So far I do really like this. I was worried about avatar images being stored on the user objects, but at these sizes I don't think they will cause many problems.

Parse recommends the use of Parse.File to store media:
http://parseplatform.github.io/docs/js/guide/#files
If for any reason we need to send in larger data we should look into that instead.

Some other feedback:
- Due to PixelRatio, the profile images are now looking very stretched and pixelated on the ProfilePage. Images are stored in 180px, that is not the same as the 180 unit we set in the StyleSheet.
- No feedback to a user if the image fails to upload. Maybe a simple alert would do it.
- If an image fails to upload, it will be replaced in the ProfilePage, but not in the menu. The ProfilePage picture goes back to the previous one when restarting the app.

File: js/router/SharedNavigator.js [hunk 1]
```diff
@@ -238,7 +235,7 @@ const SharedNavigator = React.createClass({
         viewComponent = <RegistrationPages {...sharedProps} index={route.index} />;
         break;
       case 'profile':
-        viewComponent = <ProfilePage {...sharedProps} />;
+        viewComponent = <ProfilePage {...sharedProps} controPanel={this._controlPanel} />;
```
Comment (sophiabersani) [inline js/router/SharedNavigator.js]: This seems to be unused?

File: ios/GoodQuestion/AppDelegate.m [hunk 2]
```diff
@@ -32,7 +32,7 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
    * on the same Wi-Fi network.
    */
 
-  jsCodeLocation = [NSURL URLWithString:@"http://localhost:8081/index.ios.bundle?platform=ios&dev=true"];
+  jsCodeLocation = [NSURL URLWithString:@"http://192.168.1.41:8081/index.ios.bundle?platform=ios&dev=true"];
```
Comment (sophiabersani) [inline ios/GoodQuestion/AppDelegate.m]: Left-over hardcoded local IP.

**Comment by dan-nyanko** [comment]
@gbersani 

I have tested Parse.File and something to be aware of is that the file is stored on S3.

```
file = new Parse.File();
// some missing code that would attach image
// ...
file.save();
// attache file reference to user
user.avatar(file);
user.save();
```

Inside Parse, the file reference has properties for `name` and `url`.  The url property is the location on S3.  So one would have to a) have internet access to download the image b) setup a way to store this image to asyncStorage c) sync this image upon login/logout.  Having considered this and knowing that the user record is already synchronized and cached to local storage it makes sense to simply base64 encode the image onto the profile record.  

Thanks for catching the case on failed upload, I'm assuming you mean saving the user record?  So if Parse Server is offline, calling `user.save()` fails but the Profile Page image is updated?

**Comment by sophiabersani** [comment]
Interesting, that's good to know on the Parse.File behaviour, definitely something we'll keep in mind when adding more media capabilities to the app.

Here are steps to reproduce the bug:
- Login
- Simulate a bad connection (airplane mode, disable wifi, etc)
- Attempt to submit a new picture
  - At this stage the app will not warn the user it failed
- Click the "back" button
- Go back to "Profile".
  - The page will now show the picture which failed to upload as their avatar.
  - The picture will also reset to the previous one after restarting the app.

**Comment by dan-nyanko** [comment]
@gbersani - The source of react-native-image-picker shows that the image is resized to fit the dimensions specified in the API.  You can see the source code [java](https://github.com/marcshilling/react-native-image-picker/blob/master/android/src/main/java/com/imagepicker/ImagePickerModule.java#L577) and [ios](https://github.com/marcshilling/react-native-image-picker/blob/master/ios/ImagePickerManager.m#L535). The resizing of the image is why it is lower quality. I suppose that I could make the image size 50 to 150% bigger than the component dp to increase that quality but I'm not sure its worth it.  Let me know what you think.

Also, I added an alert and rolled back the image to fix the edge case you identified. Thanks!

**Comment by sophiabersani** [comment]
@dan-nyanko The quality is lower due to PixelRatio + compression, and I believe that the avatars are still getting stretched out as well, making them look even worse.
- Pictures are being saved in a maximum of 180px h/w
- Pictures are being displayed with a minimum of 180x180

I tested it on both devices, the bigger the disparity between height/width, the lower quality the avatar will be.
eg. A 1800x900 pic will be scaled down to 180x90

You can enable cropping passing `allowsEditing` in options.
According to the docs it should work on iOS natively. It looks like it can also work on Android with some configuration. :)

**Comment by sophiabersani** [comment]
Warning is working great for failed network requests, but it also comes up when the user cancels the "Pick Photo" modal. Is this intended?

**Comment by dan-nyanko** [comment]
@gbersani 

The poor quality is from upscaling a small image (maxHeight 180, maxWidth 180) to the devices dp. Compression is applied each time a lossless format is saved, such as .jpeg or .png.

Like I previously stated, I don't feel its worth storing a 50% - 150% bigger image depending the the PixelRatio because a) it is persisted to storage b) it would be inconsistent across the users devices.  The PixelRatio is simply a [multiplier](https://facebook.github.io/react-native/docs/pixelratio.html#fetching-a-correctly-sized-image).

Cropping was considered but the tested library does not make it available on both platforms. I also considered a 3rd party service to perform auto-cropping, such as transloadit, but again did not feel that the application warranted such a service as the only images we are working with is a user avatar.

_If you have a method to allow cropping on Android as well, please let me know!_

I'll make a change to suppress the error on selecting 'Cancel'.

**Pull request merged**